pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,19 +215,97 @@ func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) [
 	return tasks
 }
 
-// SuccessfulPipelineTaskNames returns a list of the names of all of the PipelineTasks in state
-// which have successfully completed.
-func (state PipelineRunState) SuccessfulPipelineTaskNames() []string {
-	done := []string{}
+// SuccessfulOrSkippedDAGTasks returns a list of the names of all of the PipelineTasks in state
+// which have successfully completed or skipped
+func (state PipelineRunState) SuccessfulOrSkippedDAGTasks(d *dag.Graph) []string {
+	tasks := []string{}
 	for _, t := range state {
-		if t.TaskRun != nil {
-			c := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-			if c.IsTrue() {
-				done = append(done, t.PipelineTask.Name)
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.IsSuccessful() || isSkipped(t, state.ToMap(), d) {
+				tasks = append(tasks, t.PipelineTask.Name)
+			}
+		}
+	}
+	return tasks
+}
+
+// isDAGTasksStopped returns true if any of the pipeline task has failed
+// and none of the pipleine task are still running
+func (state PipelineRunState) isDAGTasksStopped(d *dag.Graph) bool {
+	failed := false
+	for _, t := range state {
+		if t.IsFailure() {
+			failed = true
+			continue
+		}
+		if t.IsStarted() && !t.IsDone() {
+			failed = false
+			break
+		}
+	}
+	return failed
+}
+
+// isDAGTasksDone returns true if all DAG tasks are done executing,
+// DAG tasks could have failed/succeeded/skipped
+func (state PipelineRunState) isDAGTasksDone(d *dag.Graph) (isDone bool) {
```

Comment:
NIT: `is` with plural `DAGTasks` feels a bit strange. Perhaps `isDAGDone` ?

Reply:
I would like to keep it as is since the helping verb `is` is referring to the verb `Done` 🤪

Renaming it to `isDAGDone` would also mean changing it to `isFinalDone` which sounds odd too.

Or  May be change it to `CheckDAGTasksDone`, `CheckFinalTasksDone`? (too verbose)

Reply:
> I would like to keep it as is since the helping verb `is` is referring to the verb `Done` 🤪

uh?

> 
> Renaming it to `isDAGDone` would also mean changing `isFinallyTasksDone` to `isFinalDone` which sounds odd too.
> 
Heh, well, rather than renaming `isFinalTasksDone` I would remove it.
`isFinalTasksDone` is almost identical to `isDAGTasksDone`, the only difference is that `isDAGTasksDone` also checks for skipped tasks, but since tasks in the finally sections cannot use `runAfter` and `conditionals`, the extra check would not have any impact on the finally DAG. It would be safe to use a single `isDAGDone` and pass the pipeline DAG or final DAG as input as needed.

> Or May be change it to `CheckDAGTasksDone`, `CheckFinalTasksDone`? (too verbose)

Reply:
yup that was the key difference, if you think we wouldn't be wasting cycles checking for `skipped` tasks even when we know there should not be any, I can create just one function `CheckTasksDone` and remove `isFinalTasksDone`, thoughts?

Reply:
> I can create just one function CheckTasksDone and remove isFinalTasksDone, thoughts?

yaaaaassss :D

Reply:
I have created an issue #2852 to track this simplification post this PR